desktop/src/preferences/write.rs

```diff
@@ -51,127 +51,251 @@ impl<'a> PreferencesWriter<'a> {
-        let mut preferences = parse(original);
-        let mut writer = PreferencesWriter::new(&mut preferences);
-        fun(&mut writer);
-        check_roundtrip(&mut preferences);
-        assert_eq!(expected, preferences.toml_document.to_string());
-    }
+    fn get_underlying_table(&mut self) -> &mut ArrayOfTables {
+        if self.0.toml_document.contains_array_of_tables("bookmark") {
+            return self.0.toml_document["bookmark"]
+                .as_array_of_tables_mut()
+                .expect("type was just verified");
+        }
 
-    #[test]
-    fn set_graphics_backend() {
-        test(
-            "",
-            |writer| writer.set_graphics_backend(GraphicsBackend::Default),
-            "graphics_backend = \"default\"\n",
-        );
+        // Type was incorrect, time to destroy.
+        self.0.toml_document.insert("bookmark", array());
+        self.0.toml_document["bookmark"]
+            .as_array_of_tables_mut()
+            .expect("type was just created")
+    }
 
-        test(
-            "graphics_backend = \"fast\"",
-            |writer| writer.set_graphics_backend(GraphicsBackend::Vulkan),
-            "graphics_backend = \"vulkan\"\n",
-        );
+    pub fn add(&mut self, bookmark: Bookmark) {
+        let table = self.get_underlying_table();
+        // TODO: If we add a BookmarkWriter use this here instead rather than duplicating the table write code.
+        let mut bookmark_table = Table::new();
+        bookmark_table["url"] = value(bookmark.url.to_string());
+        table.push(bookmark_table);
+        self.0.values.push(bookmark);
     }
 
-    #[test]
-    fn set_graphics_power_preference() {
-        test(
-            "",
-            |writer| writer.set_graphics_power_preference(PowerPreference::High),
-            "graphics_power_preference = \"high\"\n",
-        );
+    pub fn set_url(&mut self, index: usize, url: url::Url) {
+        let table = self.bookmark_table(index);
+        table["url"] = value(url.as_str());
+        self.0.values[index].url = url;
+    }
 
-        test(
-            "graphics_power_preference = \"fast\"",
-            |writer| writer.set_graphics_power_preference(PowerPreference::Low),
-            "graphics_power_preference = \"low\"\n",
-        );
+    pub fn remove(&mut self, index: usize) {
+        let table = self.get_underlying_table();
+        table.remove(index);
+        self.0.values.remove(index);
     }
+}
 
-    #[test]
-    fn set_language() {
-        test(
-            "",
-            |writer| writer.set_language(langid!("en-US")),
-            "language = \"en-US\"\n",
-        );
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use fluent_templates::loader::langid;
 
-        test(
-            "language = \"???\"",
-            |writer| writer.set_language(langid!("en-Latn-US-valencia")),
-            "language = \"en-Latn-US-valencia\"\n",
-        );
-    }
+    macro_rules! define_test_helpers {
```

Comment:
Maybe `define_serialization_test_helpers` or something similar; "test" is a little too generic in a codebase that really should have lots :D